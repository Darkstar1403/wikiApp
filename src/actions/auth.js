import Swal from 'sweetalert2';
import { createUserWithEmailAndPassword, getAuth, signInWithEmailAndPassword, signInWithPopup, signOut, updateProfile } from 'firebase/auth';
import { googleAuthProvider } from '../components/firebase/firebase-config';
import {types} from '../types/types';
import { finishLoading, startLoading } from './ui';
import { noteLogout } from './notes';

export const startLoginEmailPassword = (email, password) =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        return signInWithEmailAndPassword(auth, email, password)
        .then(({user})=>{
            if(user.photoURL){
                console.log(user.photoURL);
            }
            (user.photoURL)?dispatch(login(user.uid, user.displayName, true)):dispatch(login(user.uid, user.displayName));
            dispatch(finishLoading());
            Swal.fire(
                "Has ingresado a la mejor wiki",
                'Presionaste el botón',
                'success'
              );
        }).catch(e =>{
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        });
        
    }
}

export const  startRegister = (email, password, name, admin) =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        let url = null;
        if(admin){
            url = 'admin'
        }
        createUserWithEmailAndPassword(auth, email, password).then(async ({user})=>{
            await updateProfile(user, {displayName:name, photoURL: url});
            dispatch(login(user.uid, user.displayName, (url)?true:false));
            dispatch(finishLoading());
            Swal.fire(
                "Te has registrado en la mejor wiki",
                'Presionaste el botón',
                'success'
              );
        }).catch(e =>{
            console.log(e);
            dispatch(finishLoading());
            Swal.fire('Error', e.message, 'error');
        })
        
    }
}

export const startGoogleLogin = () =>{
    return (dispatch) =>{
        dispatch(startLoading());
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
            .then(({user}) =>{
                dispatch(login(user.uid, user.displayName))
            }).catch(e => console.log(e));
            dispatch(finishLoading());
        
    }
}

export const login = (uid, displayName, admin = false) =>(
    {
        type:types.login,
        payload: {
            uid,
            displayName,
            admin
        }
    }
)

export const startLogout = () =>{
    return async(dispatch) =>{
        const auth = getAuth();
       await signOut(auth);
       dispatch(logout());
       dispatch(noteLogout());
    }
}

export const logout = () =>({
    type:types.logout
})