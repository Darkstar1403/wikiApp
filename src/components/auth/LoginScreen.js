import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { Link } from "react-router-dom";
import Swal from "sweetalert2";
import { startGoogleLogin, startLoginEmailPassword } from "../../actions/auth";
import { removeError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";

export const LoginScreen = () => {

  const { loading} = useSelector(state => state.ui);

  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({email: '', password: ''})

  const {email, password} = formValues;

  const handleLogin = (e) =>{
      e.preventDefault();
      if(isFormValid()){
        dispatch(startLoginEmailPassword(email, password));
      }
  }

  const handleGoogleLogin = () =>{
    dispatch(startGoogleLogin());
  }

  const isFormValid = () =>{
    if(!/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)+$/.test(email)){
      Swal.fire('Error', 'Email is not valid', 'error');
      return false;
    }
    else if(!/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$/.test(password)){
      Swal.fire('Error', 'Password is not valid', 'error');
      return false
    }
    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className='auth__title'>Login</h3>
      <form onSubmit={handleLogin} className='animate__animated animate__fadeIn animate__faster'>
        {/*msgError && (<div className='auth__alert-error'>{msgError}</div>)*/}
        <input className='auth__input' type="email" placeholder="Email" name="email" value={email} onChange={handleInputChange}/>
        <input className='auth__input' type="password" placeholder="Password" name="password" value={password} onChange={handleInputChange}/>
        <button disabled={loading} className='btn btn-primary btn-block' type="submit">Login</button>
        <div className='auth__social-networks'>
          <p>Ingresar con redes sociales</p>
          <div 
              className="google-btn"
              onClick={handleGoogleLogin}
              
          >
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Ingresar Google</b>
              </p>
          </div>
        </div>
        <Link className='link' to='/auth/register'>
          Crear una cuenta nueva
        </Link>
      </form>
    </>
  );
};
