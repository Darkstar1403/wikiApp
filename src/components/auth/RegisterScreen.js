import React, { useState } from "react";
import { useSelector } from "react-redux";
import { useDispatch } from "react-redux";
import { Link } from "react-router-dom";
import Swal from "sweetalert2";
import { startRegister } from "../../actions/auth";
import { removeError } from "../../actions/ui";
import { useForm } from "../../hooks/useForm";



export const RegisterScreen = () => {

  const dispatch = useDispatch();
//useSelector te trae información en tiempo real del state de
//todos tus reducers de la aplicación

  const [admin, setAdmin] = useState(false);

  const {msgError, loading} = useSelector(state => state.ui);

  const [formValues, handleInputChange] = useForm({name:'', email:'', password:'', passwordConfirm:''})
  
  const {name, email, password, passwordConfirm} = formValues;

  const handleRegister = (e) =>{
    e.preventDefault();
    if(isFormValid()){
      dispatch(startRegister(email, password, name, admin))
    }
  }

  const isFormValid = () =>{
    if(!/^[a-zA-Z]\w*$/.test(name.trim())){
      Swal.fire('Error', 'El nombre solo tiene una palabra', 'error');
      return false;
    }
    else if(!/^[a-zA-Z0-9.!#$%&’*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)+$/.test(email)){
      Swal.fire('Error', 'Email no valido', 'error');
      return false;
    }
    else if(!/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,}$/.test(password)){
      Swal.fire('Error', 'Contraseña incorrecta', 'error');
      return false
    }
    else if(password !== passwordConfirm){
      Swal.fire('Error', 'Las contraseñas no concuerdan', 'error');
      return false;
    }
    dispatch(removeError());
    return true;
  }

  return (
    <>
      <h3 className='auth__title'>Registro</h3>
      <form onSubmit={handleRegister} className='animate__animated animate__fadeIn animate__faster'>
        {msgError && (<div className='auth__alert-error'>
          {msgError}
        </div>)}
        <input className='auth__input' type="text" placeholder="Nombre" name="name" value={name} onChange={handleInputChange} />
        <input className='auth__input' type="email" placeholder="Email" name="email" value={email} onChange={handleInputChange}/>
        <input className='auth__input' type="password" placeholder="Contraseña" name="password" value={password} onChange={handleInputChange}/>
        <input className='auth__input' type="password" placeholder="Confirmación de Contraseña" name="passwordConfirm" value={passwordConfirm} onChange={handleInputChange}/>
        <input className='me-3' type="checkbox" checked={admin} onChange={()=>{console.log(admin); setAdmin(a=>!a)}} />
        Administrador
        <button disabled={loading} className='btn btn-primary btn-block mb-5 mt-3' type="submit">Registrar</button>
        <Link className='link' to='/auth/login'>
          ¿Ya estas registrado?
        </Link>
      </form>
    </>
  );
};
