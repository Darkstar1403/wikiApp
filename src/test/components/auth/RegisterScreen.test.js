import thunk from "redux-thunk";
import configureStore from 'redux-mock-store';
import { Provider } from "react-redux";
import { mount } from "enzyme";
import { MemoryRouter } from "react-router";
import { RegisterScreen } from "../../../components/auth/RegisterScreen";


const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {
    auth:{},
    ui:{
        loading: false,
        msgError: null
    }
};
let store = mockStore(initState);
//store.dispatch = jest.fn();

describe('Tests of <RegisterScreen/>', () => {
    // beforeEach(()=>{
    //     store = mockStore(initState);
    //     //si usamos algun mock es buena idea limpiarlo

    // });
    const wrapper = mount(
        <Provider store={store}>
            <MemoryRouter>
                <RegisterScreen/>
            </MemoryRouter>
        </Provider>
    );

    test('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should use the correct dispatch', () => {
        const nameField = wrapper.find('input[name="name"]');

        nameField.simulate('change',{
            target:{
                value:'brancoasddd@ sdd',
                name:'name'
            }
        });
        wrapper.find('form').simulate('submit',{
            preventDefault(){}
        });
        const actions = store.getActions();
        console.log(actions);
    })
    


    
})
