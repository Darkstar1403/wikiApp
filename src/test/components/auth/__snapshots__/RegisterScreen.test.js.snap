// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of <RegisterScreen/> should render correctly 1`] = `
Array [
  <h3
    className="auth__title"
  >
    Register
  </h3>,
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      type="text"
      value=""
    />
    <input
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="email"
      value=""
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      className="auth__input"
      name="passwordConfirm"
      onChange={[Function]}
      placeholder="Confirm password"
      type="password"
      value=""
    />
    <button
      className="btn btn-primary btn-block mb-5"
      disabled={false}
      type="submit"
    >
      Register
    </button>
    <a
      className="link"
      href="/auth/login"
      onClick={[Function]}
    >
      Already registered?
    </a>
  </form>,
]
`;
